src/Shared/packet_adapter/packet_parser/infra/packet_parser.go

```diff
@@ -1,142 +1,141 @@
-	for _, measurement := range parser.packetTypes[id] {
-		if measurement.Name == name {
-			return measurement
-		}
-	}
-	panic("measurement not found")
-}
+package infra
+
+import (
+	"bytes"
+	"io"
+	"log"
+	"strconv"
+
+	excelAdapter "github.com/HyperloopUPV-H8/Backend-H8/Shared/excel_adapter/domain"
+	"github.com/HyperloopUPV-H8/Backend-H8/Shared/packet_adapter/packet_parser/domain"
+	"github.com/HyperloopUPV-H8/Backend-H8/Shared/packet_adapter/packet_parser/infra/dto"
+	"github.com/HyperloopUPV-H8/Backend-H8/Shared/packet_adapter/packet_parser/infra/serde"
+)
+
+type id = uint16
+type packetMeasurements = []domain.MeasurementData
+type name = string
+
+type PacketParser struct {
+	packetTypes map[id]packetMeasurements
+	enums       map[name]domain.Enum
+}
+
+func NewParser(packets []excelAdapter.PacketDTO) PacketParser {
+	return PacketParser{
+		packetTypes: getPacketTypes(packets),
+		enums:       getEnums(packets),
+	}
+}
+
+func getEnums(packets []excelAdapter.PacketDTO) map[name]domain.Enum {
+	enums := make(map[name]domain.Enum, 0)
+	for _, packet := range packets {
+		for _, measurement := range packet.Measurements {
+			if domain.IsEnum(measurement.ValueType) {
+				enums[measurement.Name] = domain.NewEnum(measurement.ValueType)
+			}
+		}
+	}
+
+	return enums
+}
+
+func getPacketTypes(packets []excelAdapter.PacketDTO) map[uint16]packetMeasurements {
+	packetMeasurements := make(map[id]packetMeasurements, len(packets))
+
+	for _, packet := range packets {
+		measurementDataArr := getMeasurementData(packet)
+		id, err := strconv.ParseUint(packet.Description.ID, 10, 16)
+
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		packetMeasurements[uint16(id)] = measurementDataArr
+	}
+
+	return packetMeasurements
+}
+
+func getMeasurementData(packet excelAdapter.PacketDTO) []domain.MeasurementData {
+	measurementDataArr := make([]domain.MeasurementData, len(packet.Measurements))
+	for index, measurement := range packet.Measurements {
+		valueType := measurement.ValueType
+		if domain.IsEnum(valueType) {
+			valueType = "enum"
+		}
+		measurementData := domain.NewMeasurement(measurement.Name, valueType)
+		measurementDataArr[index] = measurementData
+	}
+	return measurementDataArr
+}
+
+func (parser PacketParser) Decode(data []byte) dto.PacketUpdate {
+	dataReader := bytes.NewBuffer(data)
+	id := serde.DecodeID(dataReader)
+	values := parser.decodePacket(parser.packetTypes[id], dataReader)
+
+	return dto.NewPacketUpdate(id, values, data)
+}
+
+func (parser PacketParser) decodePacket(measurements packetMeasurements, bytes io.Reader) map[name]any {
+	values := make(map[name]any, len(measurements))
+	for _, measurementData := range measurements {
+		values[measurementData.Name()] = parser.decodeMeasurement(measurementData, bytes)
+	}
+	return values
+}
+
+func (parser PacketParser) decodeMeasurement(measurement domain.MeasurementData, reader io.Reader) any {

```

Comment:
the thing is this method might return any value because it decodes to a value that we dont know the type at compile time (it might be float, string or bool right now). We couldn't find a proper solution when we wrote this, but please let us know if you can think of any other way